File: sdks/node/src/resources/teams.ts
/**
 * HookSniff SDK — Teams Resource
 */

import { HttpMethod, HookSniffRequest, type HookSniffRequestContext } from "../request";
import type {
  Team,
  TeamDetailResponse,
  CreateTeamRequest,
  InviteRequest,
  ChangeRoleRequest,
} from "../models";

export class Teams {
  constructor(private readonly requestCtx: HookSniffRequestContext) {}

  /** List all teams. */
  public list(): Promise<Team[]> {
    const request = new HookSniffRequest(HttpMethod.GET, "/v1/teams");
    return request.send(this.requestCtx, (json) => json as Team[]);
  }

  /** Get a team by ID (includes members and invites). */
  public get(teamId: string): Promise<TeamDetailResponse> {
    const request = new HookSniffRequest(HttpMethod.GET, "/v1/teams/{id}");
    request.setPathParam("id", teamId);
    return request.send(this.requestCtx, (json) => json as TeamDetailResponse);
  }

  /** Create a new team. */
  public create(body: CreateTeamRequest): Promise<Team> {
    const request = new HookSniffRequest(HttpMethod.POST, "/v1/teams");
    request.setBody(body);
    return request.send(this.requestCtx, (json) => json as Team);
  }

  /** Delete a team. */
  public delete(teamId: string): Promise<void> {
    const request = new HookSniffRequest(HttpMethod.DELETE, "/v1/teams/{id}");
    request.setPathParam("id", teamId);
    return request.sendNoResponseBody(this.requestCtx);
  }

  /** Invite a member to the team. */
  public invite(teamId: string, body: InviteRequest): Promise<void> {
    const request = new HookSniffRequest(HttpMethod.POST, "/v1/teams/{id}/invite");
    request.setPathParam("id", teamId);
    request.setBody(body);
    return request.sendNoResponseBody(this.requestCtx);
  }

  /** Remove a member from the team. */
  public removeMember(teamId: string, userId: string): Promise<void> {
    const request = new HookSniffRequest(HttpMethod.DELETE, "/v1/teams/{id}/members/{uid}");
    request.setPathParam("id", teamId);
    request.setPathParam("uid", userId);
    return request.sendNoResponseBody(this.requestCtx);
  }

  /** Change a member's role. */
  public changeMemberRole(teamId: string, userId: string, body: ChangeRoleRequest): Promise<void> {
    const request = new HookSniffRequest(HttpMethod.PUT, "/v1/teams/{id}/members/{uid}/role");
    request.setPathParam("id", teamId);
    request.setPathParam("uid", userId);
    request.setBody(body);
    return request.sendNoResponseBody(this.requestCtx);
  }
}

File: sdks/rust/src/resources/teams.rs
//! HookSniff API Resource: Teams
//!
//! Team management — list members, invite, remove.

use crate::request::{HookSniffRequest, HookSniffRequestContext, HttpMethod};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct TeamMember {
    pub id: String,
    pub email: String,
    pub name: String,
    pub role: String,
    pub joined_at: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct InviteInput {
    pub email: String,
    pub role: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct InviteOutput {
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub invite_id: Option<String>,
}

#[derive(Debug)]
pub struct Teams {
    ctx: HookSniffRequestContext,
}

impl Teams {
    pub fn new(ctx: HookSniffRequestContext) -> Self {
        Self { ctx }
    }

    /// List team members
    pub fn members(&self) -> Result<Vec<TeamMember>, Box<dyn std::error::Error>> {
        let req = HookSniffRequest::new(HttpMethod::Get, "/v1/teams/members");
        req.send(&self.ctx)
    }

    /// Invite a team member
    pub fn invite(&self, input: &InviteInput) -> Result<InviteOutput, Box<dyn std::error::Error>> {
        let mut req = HookSniffRequest::new(HttpMethod::Post, "/v1/teams/invite");
        req.set_body(input);
        req.send(&self.ctx)
    }

    /// Remove a team member
    pub fn remove_member(&self, id: &str) -> Result<(), Box<dyn std::error::Error>> {
        let mut req = HookSniffRequest::new(HttpMethod::Delete, "/v1/teams/members/{id}");
        req.set_path_param("id", id);
        req.send_void(&self.ctx)
    }
}
